File: searching-algorithms/interpolation_search.js
"use strict";

const interpolationSearch = (searchNumber, inputArray) => {
  try {
    for (let i = 0; i < inputArray.length; i++) {
      if (i == 0) {
        continue;
      } else {
        if (inputArray[i - 1] > inputArray[i]) {
          throw "Array is not sorted in ascending order";
        }
      }
    }

    let numberIndex = -1;
    let startIndex = 0;
    let endIndex = inputArray.length - 1;

    while (startIndex <= endIndex) {
      let searchIndex = Math.floor(
        ((endIndex - startIndex) * (searchNumber - inputArray[startIndex])) / (inputArray[endIndex] - inputArray[startIndex])
      );

      if (searchNumber == inputArray[searchIndex]) {
        numberIndex = searchIndex;
        break;
      } else if (searchNumber > inputArray[searchIndex]) {
        startIndex = searchIndex + 1;
        endIndex = inputArray.length;
      } else {
        startIndex = 0;
        endIndex = searchIndex - 1;
      }
    }

    return numberIndex;
  } catch (err) {
    console.log("interpolationSearch error : ", err);
  }
};

let inputArray = [1, 3, 4, 11, 13];
let searchNumber = 13;

let numberIndex = interpolationSearch(searchNumber, inputArray);
console.log("numberIndex : ", numberIndex);

File: searching-algorithms/interpolation_search.py
import math

def is_list_sorted(input_list):
    sorted_lst = True
    for idx, each_item in enumerate(input_list):
        if idx == 0:
            continue
        else:
            current_item = each_item
            previous_item = input_list[idx - 1]
            if current_item < previous_item:
                sorted_lst = False
                break

    return sorted_lst

def run_interpolation_search(search_number, input_list):

    if not is_list_sorted(input_list):
        raise Exception("List is not sorted in ascending order")
    
    number_index = -1
    low_index = 0
    high_index = len(input_list) - 1

    while low_index <= high_index:
        prob_index = int((high_index - low_index) * (search_number - input_list[low_index]) / (input_list[high_index] - input_list[low_index]))

        if search_number == input_list[prob_index]:
            number_index = prob_index
            break
        elif search_number > input_list[prob_index]:
            low_index = prob_index + 1
            high_index = len(input_list) -1 
        else:
            high_index = prob_index - 1
            low_index = 0             

    return number_index


if __name__ == "__main__":
    search_number = 11
    input_list = [3, 4, 10, 11]
    number_index = run_interpolation_search(search_number, input_list)
    print("number found in index : ", number_index)
